Support updating the gamepad mapping.

divVerent opened the pull request:
This is only supported on desktops yet (on mobile standard layout isn't
implemented yet, and on the web this is the browser's responsibility).

Fixes #1723.

divVerent commented:
FYI - not tested yet; for now just sent for early review. Going to test early next week.

hajimehoshi commented on input.go:
```diff
@@ -234,6 +234,15 @@ func IsStandardGamepadLayoutAvailable(id GamepadID) bool {
+//
+// This functionality is only supported on desktop systems and will just return an error elsewhere.
+//
+// UpdateStandardGamepadLayoutMappings must be called on the main thread before the main loop.
```
Is this true? We should be able to make this concurrent-safe.

divVerent replied:
Done now (by moving it to the ui.t thread as many other places do).

This doesn't seem to be possible BEFORE the main loop is launched, so before the main loop it remains concurrent-unsafe.

hajimehoshi commented on input.go:
```diff
@@ -234,6 +234,15 @@ func IsStandardGamepadLayoutAvailable(id GamepadID) bool {
+//
+// UpdateStandardGamepadLayoutMappings must be called on the main thread before the main loop.
+func UpdateStandardGamepadLayoutMappings(mappings string) error {
+	return uiDriver().Input().UpdateStandardGamepadLayoutMappings(mappings)
```
Just returning a boolean value should be fine. As some platforms (js) don't support it on purpose, this should not be an 'error'.

hajimehoshi reviewed (commented):
Thank you!

divVerent commented:
On Sat, Aug 21, 2021, 02:14 Hoshi, Hajime (星一) ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Thank you!
> ------------------------------
>
> In input.go
> <https://github.com/hajimehoshi/ebiten/pull/1775#discussion_r693310717>:
>
> > @@ -234,6 +234,15 @@ func IsStandardGamepadLayoutAvailable(id GamepadID) bool {
>  	return uiDriver().Input().IsStandardGamepadLayoutAvailable(id)
>  }
>
> +// UpdateStandardGamepadLayoutMappings updates the gamepad layout definitions using a set of definitions in SDL_GameControllerDB format.
> +//
> +// This functionality is only supported on desktop systems and will just return an error elsewhere.
> +//
> +// UpdateStandardGamepadLayoutMappings must be called on the main thread before the main loop.
>
> Is this true? We should be able to make this concurrent-safe.
>
GLFW docs say it must be run on the main thread. I could do the dance to
run it on the main thread from anywhere like many other functions do - did
not see it as necessary though.

What I do not yet know is when mapping apply when you do this when the game
is already running - e.g. does one need to reconnect the gamepad to
actually apply changes to it.

Will try, then do this change and document it properly.



------------------------------
>
> In input.go
> <https://github.com/hajimehoshi/ebiten/pull/1775#discussion_r693311425>:
>
> > @@ -234,6 +234,15 @@ func IsStandardGamepadLayoutAvailable(id GamepadID) bool {
>  	return uiDriver().Input().IsStandardGamepadLayoutAvailable(id)
>  }
>
> +// UpdateStandardGamepadLayoutMappings updates the gamepad layout definitions using a set of definitions in SDL_GameControllerDB format.
> +//
> +// This functionality is only supported on desktop systems and will just return an error elsewhere.
> +//
> +// UpdateStandardGamepadLayoutMappings must be called on the main thread before the main loop.
> +func UpdateStandardGamepadLayoutMappings(mappings string) error {
> +	return uiDriver().Input().UpdateStandardGamepadLayoutMappings(mappings)
>
> Just returning a boolean value should be fine. As some platforms (js)
> don't support it on purpose, this should not be an 'error'.
>
Will do; the reason why I went with the error is to distinguish the two
reasons "not needed here" and "failed to parse".

So to fix this, are you OK if I make it return false on mobile (not
implemeted yet), true on JS (not needed) and forward the value on desktop
(i.e. indicating parse error)?

Or would you prefer true on mobile too where standard layout is not
implemented at all yet?


—
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hajimehoshi/ebiten/pull/1775#pullrequestreview-735470964>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB5NMDXY643GIFYCBQOYN3T5474FANCNFSM5CRCRDVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

hajimehoshi commented:
> What I do not yet know is when mapping apply when you do this when the game
> is already running - e.g. does one need to reconnect the gamepad to
> actually apply changes to it.
> Will try, then do this change and document it properly.

Ah good point. I appreciate your trial.

> So to fix this, are you OK if I make it return false on mobile (not
> implemeted yet), true on JS (not needed) and forward the value on desktop
> (i.e. indicating parse error)?

Hmm, I see. What about returning `(bool, error)`? The mobile will always return `false, nil`.

hajimehoshi commented on input.go:
```diff
@@ -234,6 +234,18 @@ func IsStandardGamepadLayoutAvailable(id GamepadID) bool {
 
+// UpdateStandardGamepadLayoutMappings updates the gamepad layout definitions using a set of definitions in SDL_GameControllerDB format.
+//
+// Returns whether the mappings were applied, and whether an error occurred parsing the mappings.
```
* `This reports whether ...` or `UpdateStandardGamepadLayoutMappings reports whether ...`
* `and this returns an error occurred in parsing the mappings if it exists`

divVerent replied:
First applied. Sorry, I do not understand your second suggestion; rephrased another way, is this clearer?

hajimehoshi replied:
The latest version seems good, thanks!

hajimehoshi commented on input.go:
```diff
@@ -234,6 +234,18 @@ func IsStandardGamepadLayoutAvailable(id GamepadID) bool {
+//
+// Returns whether the mappings were applied, and whether an error occurred parsing the mappings.
+//
+// On platforms where gamepad mappings are not managed by ebiten, this always returns false, nil.
```
`false and nil`

divVerent replied:
Done.

hajimehoshi commented on input.go:
```diff
@@ -234,6 +234,18 @@ func IsStandardGamepadLayoutAvailable(id GamepadID) bool {
+//
+// On platforms where gamepad mappings are not managed by ebiten, this always returns false, nil.
+//
+// UpdateStandardGamepadLayoutMappings must be called on the main thread before ebiten.Run, and is concurrent-safe after ebiten.Run.
```
Update this comment.

divVerent replied:
Yes; now it's correct: before ebiten.Run it must be the main thread; after the game loop is running, it's concurrent-safe.

hajimehoshi replied:
Thanks! Perhaps the other functions might be in the same situation, though their comments say just 'concurrent-safe'. I'll revisit them...

hajimehoshi commented on input.go:
```diff
@@ -234,6 +234,18 @@ func IsStandardGamepadLayoutAvailable(id GamepadID) bool {
+// On platforms where gamepad mappings are not managed by ebiten, this always returns false, nil.
+//
+// UpdateStandardGamepadLayoutMappings must be called on the main thread before ebiten.Run, and is concurrent-safe after ebiten.Run.
+// Updated mappings take effect immediately even for already connected gamepads.
```
Insert an empty line above this.

divVerent replied:
Done.

hajimehoshi commented on internal/uidriver/mobile/input.go:
```diff
@@ -18,6 +18,8 @@
 package mobile
 
 import (
+	"errors"
```
This is no longer needed

divVerent replied:
Oops, fixed.

hajimehoshi commented on internal/uidriver/js/input_js.go:
```diff
@@ -16,6 +16,7 @@ package js
 
 import (
 	"encoding/hex"
+	"errors"
```
This is no longer needed.

divVerent replied:
Done. Oops.

hajimehoshi commented on internal/uidriver/glfw/input.go:
```diff
@@ -452,3 +453,24 @@ func standardButtonToGLFWButton(button driver.StandardGamepadButton) glfw.Gamepa
+
+func (i *Input) updateStandardGamepadLayoutMappings(mapping string) error {
+	if !glfw.UpdateGamepadMappings(mapping) {
+		return errors.New("glfw: could not parse or update gamepad mappings")
```
Optional: `fmt.Errorf`

divVerent replied:
Done. Also added a comment explaining why it's a constant error string for now (go-gl's glfw API does not provide the error, that's why).

divVerent commented:
BTW should I leave out the "take effect immediately" comment? It is true for GLFW according to its source, but I can neither tell if it's gonna stay this way in GLFW in the future, nor if the not yet written mobile support will have this property too.

Very likely semantics possible everywhere is applying it to all gamepads if run before the main loop, and any newly connected gamepads when run inside the main loop.

divVerent commented:
Another thing is - will the mobile support use the same format, or would you prefer a differently formatted string when implementing this for mobile? Could make the exact format of the string explained as platform dependent instead, and just say that on desktop platforms it's gonna be a SDL_GameControllerDB string.

hajimehoshi commented:
Thank you for updating!

> BTW should I leave out the "take effect immediately" comment? It is true for GLFW according to its source, but I can neither tell if it's gonna stay this way in GLFW in the future (as the documentation nowhere states this), nor if the not yet written mobile support will have this property too.
> Very likely semantics possible everywhere is applying it to all gamepads if run before the main loop, and any newly connected gamepads when run inside the main loop.

Hmm, let's keep the comment and the current behavior for usability. We have to be careful about GLFW's update. And we should emulate exactly the same behavior for mobiles.

BTW, does "take effect immediately" mean that the mapping for an already-connected gamepad is also updated?

> Another thing is - will the mobile support use the same format, or would you prefer a differently formatted string when implementing this for mobile? Could make the exact format of the string explained as platform dependent instead, and just say that on desktop platforms it's gonna be a SDL_GameControllerDB string.

I plan to use exactly same format for mobiles. Actually the database already includes Android and iOS data. I don't want to bother Ebiten users with forcing different formats for other platforms.

divVerent commented:
> BTW, does "take effect immediately" mean that the mapping for an already-connected gamepad is also updated?

Yes; just confirmed this. Moved my code to update gamepad handling in the jump key handling, and provided a mapping with left/right directions swapped. Was hilarious, on the first jump the directions reversed.

So GLFW does properly apply this.

Does BTW mean that when wanting to add multiple mappings, one better adds them with one call than with many call - otherwise this becomes O(n^2) as every time this function is called, all mappings are evaluated for all connected gamepads.

divVerent commented:
> FYI - not tested yet; for now just sent for early review. Going to test early next week.

Change now tested properly and working. Ready for full review / merge now.

divVerent commented:
js build failure fixed, thanks for the automated workflows!

divVerent commented:
Sorry, there was another js failure. Fixed too now.

hajimehoshi commented:
Thanks!

> Does BTW mean that when wanting to add multiple mappings, one better adds them with one call than with many call - otherwise this becomes O(n^2) as every time this function is called, all mappings are evaluated for all connected gamepads.

Hm, I don't think this function is so often called. Then I think we don't have to care about the computational cost.

divVerent commented:
Yeah - just wondering if I should explicitly state in the comment that
multiple mappings separated by newlines are permitted.

On Sun, Aug 22, 2021, 10:34 Hoshi, Hajime (星一) ***@***.***>
wrote:

> Thanks!
>
> Does BTW mean that when wanting to add multiple mappings, one better adds
> them with one call than with many call - otherwise this becomes O(n^2) as
> every time this function is called, all mappings are evaluated for all
> connected gamepads.
>
> Hm, I don't think this function is so often called. Then I think we don't
> have to care about the computational cost.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hajimehoshi/ebiten/pull/1775#issuecomment-903278396>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB5NMBYEYMBXJRNJZWH4H3T6EDHHANCNFSM5CRCRDVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

hajimehoshi commented:
> Yeah - just wondering if I should explicitly state in the comment that
> multiple mappings separated by newlines are permitted.

We should permit this explicitly. Please fix the comment. Thanks!

hajimehoshi commented on input.go:
```diff
@@ -234,6 +234,26 @@ func IsStandardGamepadLayoutAvailable(id GamepadID) bool {
+// Multiple input definitions can be provided separated by newlines.
+// In particular, it is valid to pass an entire gamecontrollerdb.txt file.
+//
+// Note though that ebiten already includes its own copy of this file,
```
s/ebiten/Ebiten/

divVerent replied:
Done.

hajimehoshi commented on input.go:
```diff
@@ -234,6 +234,26 @@ func IsStandardGamepadLayoutAvailable(id GamepadID) bool {
+//
+// UpdateStandardGamepadLayoutMappings reports whether the mappings were applied, and returns an error in case any occurred while parsing the mappings.
+//
+// On platforms where gamepad mappings are not managed by ebiten, this always returns false and nil.
```
ditto

divVerent replied:
Done.

divVerent commented:
I expanded upon the comments to describe this, and also now explain that it is valid but not necessary to pass an entire gamecontrollerdb.txt file, that this feature ideally should be exposed to the user in some form, and that people who have hardware needing this facility should send a pull request to have it included in the official database.

hajimehoshi reviewed (approved):
LGTM, thanks!
